### PR TITLE
Process stats dir refactor and enhancements

### PR DESCRIPTION
**Pull request by graydon**
Factors some of the code for reading stats-dirs out of process-stats-dir.py so it can be used by other scripts. Also adds a couple new features for grouping, selecting, and immediate comparison to process-stats-dir.py (steps towards satisfying user feedback as in rdar://33973414)

### CODE REVIEW

**Comment by graydon** [comment]
@swift-ci please test

**Comment by graydon** [comment]
@swift-ci please test

File: utils/jobstats/jobstats.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+import datetime
+import itertools
+import json
+import os
+import random
+import re
+
+
+class JobStats:
```
Comment (gottesmm) [inline utils/jobstats/jobstats.py]: Why use old style classes?

File: utils/jobstats/jobstats.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+            },
+            "Tests":
+            [
+                {
+                    "Data": [v],
+                    "Info": {},
+                    "Name": "%s.%s.%s.%s" % (args.lnt_tag, self.module,
+                                             k, self.pick_lnt_metric_suffix(k))
+                }
+                for (k, v) in stats.items()
```
Comment (gottesmm) [inline utils/jobstats/jobstats.py]: This is sort of awesome! I have never thought of doing this in large dictionary blobs!

File: utils/jobstats/jobstats.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    jobstats = []
+    auxpat = (r"(?P<module>[^-]+)-(?P<input>[^-]+)-(?P<triple>[^-]+)" +
+              r"-(?P<out>[^-]+)-(?P<opt>[^-]+)")
+    fpat = (r"^stats-(?P<start>\d+)-swift-(?P<kind>\w+)-" +
+            auxpat +
+            r"-(?P<pid>\d+)(-.*)?.json$")
+    for root, dirs, files in os.walk(path):
+        for f in files:
+            m = re.match(fpat, f)
+            if m:
```
Comment (gottesmm) [inline utils/jobstats/jobstats.py]: Why not invert the condition and save a level of indentation?

File: utils/jobstats/jobstats.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    """Loads all stats-files found in path into a list of JobStats objects"""
+    jobstats = []
+    auxpat = (r"(?P<module>[^-]+)-(?P<input>[^-]+)-(?P<triple>[^-]+)" +
+              r"-(?P<out>[^-]+)-(?P<opt>[^-]+)")
+    fpat = (r"^stats-(?P<start>\d+)-swift-(?P<kind>\w+)-" +
+            auxpat +
+            r"-(?P<pid>\d+)(-.*)?.json$")
+    for root, dirs, files in os.walk(path):
+        for f in files:
+            m = re.match(fpat, f)
```
Comment (gottesmm) [inline utils/jobstats/jobstats.py]: Why not just compile the regex?

File: utils/jobstats/jobstats.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+            if m:
+                # NB: "pid" in fpat is a random number, not unix pid.
+                mg = m.groupdict()
+                jobkind = mg['kind']
+                jobid = int(mg['pid'])
+                start_usec = int(mg['start'])
+                module = mg["module"]
+                if len(select_module) != 0 and module not in select_module:
```
Comment (gottesmm) [inline utils/jobstats/jobstats.py]: I think you should be able to just do:

```
if module not in select_module:
```

File: utils/jobstats/jobstats.py [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+                    continue
+                jobargs = [mg["input"], mg["triple"], mg["out"], mg["opt"]]
+
+                j = json.load(open(os.path.join(root, f)))
```
Comment (gottesmm) [inline utils/jobstats/jobstats.py]: Hmmm... shouldn't this be in a with statement? I have never seen someone do this as such.

File: utils/process-stats-dir.py [hunk 7]
```diff
@@ -415,6 +258,37 @@ def compare_to_csv_baseline(args):
...
 
 
+def compare_to_csv_baseline(args):
+    old_stats = read_stats_dict_from_csv(args.compare_to_csv_baseline)
+    sel = args.select_module
+    m = merge_all_jobstats([s for d in args.remainder
+                            for s in load_stats_dir(d, select_module=sel)],
+                           select_module=sel,
+                           group_by_module=args.group_by_module)
+    old_stats = dict([(k, v) for (k, (_, v)) in old_stats.items()])
```
Comment (gottesmm) [inline utils/process-stats-dir.py]: Can't you drop the [] here and just do:

```
dict((k, v) for (k, (_, v)) in old_stats.items())
```

File: utils/process-stats-dir.py [hunk 8]
```diff
@@ -415,6 +258,37 @@ def compare_to_csv_baseline(args):
     return regressions
 
 
+def compare_to_csv_baseline(args):
+    old_stats = read_stats_dict_from_csv(args.compare_to_csv_baseline)
+    sel = args.select_module
+    m = merge_all_jobstats([s for d in args.remainder
```
Comment (gottesmm) [inline utils/process-stats-dir.py]: Can't you drop the [] here?

File: utils/jobstats/jobstats.py (hunk 5 above)
Comment (graydon) [reply utils/jobstats/jobstats.py]: No, I only want to filter if it's set to a nonempty list. If it's an empty list, I want to interpret that as "select everything"

File: utils/jobstats/jobstats.py (hunk 1 above)
Comment (graydon) [reply utils/jobstats/jobstats.py]: I don't know about new-style classes. Also I have no idea which python any given style of classes is compatible with. I figured 'class X' has got to be the most widely used?

File: utils/jobstats/jobstats.py (hunk 4 above)
Comment (graydon) [reply utils/jobstats/jobstats.py]: sure

File: utils/jobstats/jobstats.py (hunk 3 above)
Comment (graydon) [reply utils/jobstats/jobstats.py]: ok

File: utils/jobstats/jobstats.py (hunk 6 above)
Comment (graydon) [reply utils/jobstats/jobstats.py]: sure

File: utils/process-stats-dir.py (hunk 8 above)
Comment (graydon) [reply utils/process-stats-dir.py]: yeah, huon keeps trying to teach me to do that :)

File: utils/process-stats-dir.py (hunk 7 above)
Comment (graydon) [reply utils/process-stats-dir.py]: likewise

File: utils/process-stats-dir.py (hunk 8 above)
Comment (graydon) [reply utils/process-stats-dir.py]: Hah, actually ...

    SyntaxError: Generator expression must be parenthesized if not sole argument

Learn something every day!

**Comment by graydon** [comment]
@swift-ci please test

File: utils/jobstats/jobstats.py (hunk 1 above)
Comment (gottesmm) [reply utils/jobstats/jobstats.py]: Info: https://wiki.python.org/moin/NewClassVsClassicClass

Just do:

```
class JobStats(object):
```

and your good.

Comment (graydon) [reply utils/jobstats/jobstats.py]: Done. Thanks!

**Comment by graydon** [comment]
@swift-ci please test and merge

**Comment by graydon** [comment]
@swift-ci please test and merge

**Pull request merged**